src/openms/include/OpenMS/KERNEL/MSExperiment.h

```diff
@@ -392,6 +392,395 @@ namespace OpenMS
+    auto getCoveredSpectra = [](
+        const std::vector<std::reference_wrapper<const MSSpectrum>>& spectra_view, 
+        std::vector<std::pair<RangeMZ, RangeRT>>& mz_rt_ranges) 
+        ->  std::vector<std::pair<size_t, size_t>>
+      {
+        std::vector<std::pair<size_t, size_t>> res;
+        res.reserve(mz_rt_ranges.size());
+        for (const auto & mz_rt : mz_rt_ranges) 
+        {
+          auto start_it = std::lower_bound(spectra_view.begin(), spectra_view.end(), mz_rt.second.getMin(), 
+            [](const auto& spec, double rt) 
+            { return spec.get().getRT() < rt; });
+
+          auto stop_it = std::upper_bound(spectra_view.begin(), spectra_view.end(), mz_rt.second.getMax(), 
+            [](double rt, const auto& spec) 
+            { return rt < spec.get().getRT(); });
+
+            res.emplace_back(
+                std::distance(spectra_view.begin(), start_it),
+                std::distance(spectra_view.begin(), stop_it)
+            );
+        }
+        return res;
+      };
+
+    // For each range, gets (spectrum start index, spectrum stop index). The spectra covered by each RT range.
+    const std::vector<std::pair<size_t, size_t>> rt_ranges_idcs = getCoveredSpectra(spectra_view, mz_rt_ranges);
+
+    // Initialize result vector
+    std::vector<MSChromatogram> result(mz_rt_ranges.size());
+
+    // Initialize counts per spectrum index and total mappings
+    std::vector<std::vector<size_t>> spec_idx_to_range_idx(spectra_view.size());
+
+    // Build spectrum to range index mapping
+    for (size_t i = 0; i < rt_ranges_idcs.size(); ++i) 
+    {
+        const auto& [start, stop] = rt_ranges_idcs[i];
+        result[i].resize(stop - start);
+        result[i].getProduct().setMZ(
+          (mz_rt_ranges[i].first.getMinMZ() + mz_rt_ranges[i].first.getMaxMZ()) / 2.0);
+        for (size_t j = start; j < stop; ++j) 
+        {
+            spec_idx_to_range_idx[j].push_back(i);
+        }
+    }
+
+   #pragma omp parallel for schedule(dynamic)
+   for (size_t i = 0; i < spec_idx_to_range_idx.size(); ++i) 
+   {
+      const auto& spec = spectra_view[i].get();
+      const double rt = spec.getRT();
+      MSSpectrum::ConstIterator spec_begin = spec.cbegin();
+      MSSpectrum::ConstIterator spec_end = spec.cend();
+
+      for (size_t range_idx : spec_idx_to_range_idx[i]) {
+        const auto& mz_range = mz_rt_ranges[range_idx].first;
+        
+        // Find data points within MZ range
+        auto start_it = spec.PosBegin(spec_begin, mz_range.getMinMZ(), spec_end);
+        auto end_it = start_it;
+        
+        while (end_it != spec_end && end_it->getPosition() < mz_range.getMaxMZ()) 
+        {
+          ++end_it;
+        }
+
+        // Calculate result using provided reduction function
+        result[range_idx][i - rt_ranges_idcs[range_idx].first] = 
+            ChromatogramPeak(rt, func_mz_reduction(start_it, end_it));
+      }
```

Comment:
looks like this is the only difference to `aggregate()`?
Can you refactor a little to save some code duplication?
Happy to help

Reply:
That would be appreciated. I did not want to use too much meta programming but maybe it is not too bad?

Reply:
Yeah return type templatization was on our list! Would be cool.

Reply:
In this case, you could actually consider it as the "aggregation function" of the second dimension (which is also something we are planning).

Reply:
and updateRanges after everything is done

Reply:
My plan was to add a second function that works on top of a vector of the result of the first one.
MzReduce = ChromPeak(SpecIt, SpecIt)
RTReduce = Chrom(vector<ChromPeak>)

Reply:
a ok like a back_inserter ?